pkg/ingester/ingester_v2.go

```diff
@@ -696,7 +719,14 @@ func (i *Ingester) v2Push(ctx context.Context, req *cortexpb.WriteRequest) (*cor
 		return nil, err
 	}
 
-	db, err := i.getOrCreateTSDB(userID, false)
+	gl := i.getGlobalLimits()
+	if gl != nil && gl.MaxIngestionRate > 0 {
```

Comment:
Can we add the new limits to `Benchmark_Ingester_v2PushOnError()` and see how it compared to master (no limit at all)? I just wanna make sure we're not missing anything from a performance perspective.

Reply:
I'm trying to do that, but I'm getting highly inconsistent results from that benchmark (both on master and in my branch). Code around it produces too much waste, and spends too much CPU on non-tested things (generating write requests, doing garbage collection).

Especially "high concurrency" scenario numbers are completely bogus. Here is what I'm getting:

```
name                                                                                            old time/op    new time/op    delta
_Ingester_v2PushOnError/failure:_out_of_bound_samples,_scenario:_no_concurrency-4                  334µs ± 3%     336µs ± 4%     ~     (p=0.365 n=20+19)
_Ingester_v2PushOnError/failure:_out_of_bound_samples,_scenario:_low_concurrency-4                12.8ms ±17%    12.6ms ± 7%     ~     (p=0.661 n=17+19)
_Ingester_v2PushOnError/failure:_out_of_bound_samples,_scenario:_high_concurrency-4                136ms ±23%     137ms ±22%     ~     (p=0.678 n=20+20)
_Ingester_v2PushOnError/failure:_out_of_order_samples,_scenario:_no_concurrency-4                  348µs ± 4%     341µs ± 2%   -2.16%  (p=0.001 n=20+20)
_Ingester_v2PushOnError/failure:_out_of_order_samples,_scenario:_low_concurrency-4                10.9ms ± 7%    10.4ms ±10%   -4.55%  (p=0.001 n=19+19)
_Ingester_v2PushOnError/failure:_out_of_order_samples,_scenario:_high_concurrency-4                111ms ± 7%     114ms ±15%     ~     (p=0.379 n=17+19)
_Ingester_v2PushOnError/failure:_per-user_series_limit_reached,_scenario:_no_concurrency-4         547µs ± 2%     563µs ± 2%   +3.01%  (p=0.000 n=20+19)
_Ingester_v2PushOnError/failure:_per-user_series_limit_reached,_scenario:_low_concurrency-4       19.1ms ± 5%    20.7ms ± 6%   +8.28%  (p=0.000 n=18+20)
_Ingester_v2PushOnError/failure:_per-user_series_limit_reached,_scenario:_high_concurrency-4       213ms ±14%     235ms ±12%  +10.39%  (p=0.000 n=20+20)
_Ingester_v2PushOnError/failure:_per-metric_series_limit_reached,_scenario:_low_concurrency-4     26.2ms ±19%    24.6ms ± 6%     ~     (p=0.588 n=20+19)
_Ingester_v2PushOnError/failure:_per-metric_series_limit_reached,_scenario:_high_concurrency-4     270ms ±16%     287ms ±38%     ~     (p=0.369 n=20+20)
_Ingester_v2PushOnError/failure:_per-metric_series_limit_reached,_scenario:_no_concurrency-4       615µs ± 2%     632µs ± 2%   +2.69%  (p=0.000 n=19+20)

name                                                                                            old alloc/op   new alloc/op   delta
_Ingester_v2PushOnError/failure:_out_of_bound_samples,_scenario:_no_concurrency-4                 99.2kB ± 0%    99.2kB ± 0%     ~     (p=0.692 n=20+19)
_Ingester_v2PushOnError/failure:_out_of_bound_samples,_scenario:_low_concurrency-4                10.5MB ± 2%    10.5MB ± 3%     ~     (p=0.841 n=20+20)
_Ingester_v2PushOnError/failure:_out_of_bound_samples,_scenario:_high_concurrency-4                127MB ±56%     136MB ±58%     ~     (p=0.583 n=20+20)
_Ingester_v2PushOnError/failure:_out_of_order_samples,_scenario:_no_concurrency-4                 2.24kB ± 2%    2.23kB ± 1%     ~     (p=0.606 n=20+20)
_Ingester_v2PushOnError/failure:_out_of_order_samples,_scenario:_low_concurrency-4                 516kB ±81%     531kB ±62%     ~     (p=0.698 n=20+20)
_Ingester_v2PushOnError/failure:_out_of_order_samples,_scenario:_high_concurrency-4              13.6MB ±275%   25.9MB ±259%  +90.42%  (p=0.033 n=19+20)
_Ingester_v2PushOnError/failure:_per-user_series_limit_reached,_scenario:_no_concurrency-4        99.3kB ± 0%    99.3kB ± 0%     ~     (p=0.273 n=19+20)
_Ingester_v2PushOnError/failure:_per-user_series_limit_reached,_scenario:_low_concurrency-4       11.1MB ± 2%    11.4MB ± 3%   +2.45%  (p=0.000 n=19+18)
_Ingester_v2PushOnError/failure:_per-user_series_limit_reached,_scenario:_high_concurrency-4       158MB ±47%     172MB ±44%     ~     (p=0.327 n=20+20)
_Ingester_v2PushOnError/failure:_per-metric_series_limit_reached,_scenario:_low_concurrency-4     11.4MB ± 9%    11.2MB ± 8%     ~     (p=0.461 n=20+20)
_Ingester_v2PushOnError/failure:_per-metric_series_limit_reached,_scenario:_high_concurrency-4     175MB ±54%     183MB ±56%     ~     (p=0.799 n=20+20)
_Ingester_v2PushOnError/failure:_per-metric_series_limit_reached,_scenario:_no_concurrency-4       100kB ± 0%     100kB ± 0%     ~     (p=0.654 n=20+20)

name                                                                                            old allocs/op  new allocs/op  delta
_Ingester_v2PushOnError/failure:_out_of_bound_samples,_scenario:_no_concurrency-4                  2.04k ± 0%     2.04k ± 0%     ~     (all equal)
_Ingester_v2PushOnError/failure:_out_of_bound_samples,_scenario:_low_concurrency-4                  206k ± 0%      206k ± 0%     ~     (p=0.799 n=20+20)
_Ingester_v2PushOnError/failure:_out_of_bound_samples,_scenario:_high_concurrency-4                2.13M ±11%     2.16M ±13%     ~     (p=0.659 n=20+20)
_Ingester_v2PushOnError/failure:_out_of_order_samples,_scenario:_no_concurrency-4                   41.0 ± 0%      41.0 ± 0%     ~     (all equal)
_Ingester_v2PushOnError/failure:_out_of_order_samples,_scenario:_low_concurrency-4                 4.94k ±29%     4.98k ±23%     ~     (p=0.733 n=20+20)
_Ingester_v2PushOnError/failure:_out_of_order_samples,_scenario:_high_concurrency-4               74.9k ±165%   115.0k ±190%  +53.45%  (p=0.026 n=19+20)
_Ingester_v2PushOnError/failure:_per-user_series_limit_reached,_scenario:_no_concurrency-4         2.03k ± 0%     2.03k ± 0%     ~     (all equal)
_Ingester_v2PushOnError/failure:_per-user_series_limit_reached,_scenario:_low_concurrency-4         207k ± 0%      208k ± 0%   +0.43%  (p=0.000 n=19+18)
_Ingester_v2PushOnError/failure:_per-user_series_limit_reached,_scenario:_high_concurrency-4       2.21M ±12%     2.26M ±11%     ~     (p=0.369 n=20+20)
_Ingester_v2PushOnError/failure:_per-metric_series_limit_reached,_scenario:_low_concurrency-4       209k ± 2%      208k ± 1%     ~     (p=0.478 n=20+20)
_Ingester_v2PushOnError/failure:_per-metric_series_limit_reached,_scenario:_high_concurrency-4     2.26M ±14%     2.28M ±15%     ~     (p=0.841 n=20+20)
_Ingester_v2PushOnError/failure:_per-metric_series_limit_reached,_scenario:_no_concurrency-4       2.04k ± 0%     2.04k ± 0%     ~     (all equal)
```

Please notice high variance in allocated memory (`±275%` !!!).

(This comparison is master vs "no limits" scenario).